src/cluster_import.c

```diff
@@ -0,0 +1,1927 @@
+     * time. */
+    struct serverCommand *cmd = lookupCommand(argv, argc);
+    getNodeByQuery(server.current_client, cmd, argv, argc, &slot, &error_code);
+    if (error_code != CLUSTER_REDIR_NONE || slot == -1) {
+        /* A couple cases where this could happen:
+         *    - The replicated command is a command without a slot.
+         *    - The replicated command is written by VM_Replicate module APIs
+         *      and is a cross-slot command, or a slot that is not owned by
+         *      this node.
+         *
+         * In any case, our best solution is to not replicate this to the
+         * target node. */
+        return;
+    }
+
+    listRewind(server.cluster->slot_migration_links, &li);
+    while ((ln = listNext(&li))) {
+        slotMigrationLink *link = (slotMigrationLink *)ln->value;
+        if (link->type != SLOT_MIGRATION_EXPORT) continue;
+        if (!link->client) continue;
+        if (!isSlotMigrationLinkInProgress(link)) continue;
+        if (!isSlotInSlotRanges(slot, link->slot_ranges)) continue;
+
+        addReplyArrayLen(link->client, argc);
+        for (i = 0; i < argc; i++) {
+            addReplyBulk(link->client, argv[i]);
+        }
+    }
+
+    if (new_argv) {
+        for (int i = 0; i < argc; i++) decrRefCount(new_argv[i]);
+        zfree(new_argv);
+    }
+}
+
+void clusterHandleSlotExportLinkClientClose(slotMigrationLink *link) {
+    link->client = NULL;
+    if (!isSlotMigrationLinkInProgress(link)) {
+        return;
+    }
+    serverLog(LL_NOTICE,
+              "Slot export connection to node %.40s (for slots [%s]) lost.",
+              link->nodename,
+              link->slot_ranges_str);
+
+    /* If we have granted failover, the failover may have happened, but we don't know. We keep the
+     * slot export around so that we remain paused until we find out about the takeover (or until
+     * the pause times out).
+     *
+     * Otherwise, we can mark it failed. */
+    if (link->state != SLOT_EXPORT_FAILOVER_GRANTED) {
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED, "Connection lost to target");
+    }
+    clusterDoBeforeSleep(CLUSTER_TODO_HANDLE_SLOT_MIGRATION);
+}
+
+int isSlotExportPauseTimedOut(slotMigrationLink *link) {
+    return link->mf_end < mstime() || !getPausedActionsWithPurpose(PAUSE_DURING_SLOT_MIGRATION);
+}
+
+void updateSlotExportIfOwnershipChanged(slotMigrationLink *link) {
+    serverAssert(link->type == SLOT_MIGRATION_EXPORT);
+    if (!isSlotMigrationLinkInProgress(link)) {
+        return;
+    }
+    int cross_node = 0;
+    clusterNode *n = getClusterNodeBySlotRanges(link->slot_ranges, &cross_node);
+    if (n) {
+        if (n == server.cluster->myself) {
+            return;
+        } else if (!memcmp(n->name, link->nodename, CLUSTER_NAMELEN)) {
+            /* All slots are now claimed by the target of this link */
+            serverLog(LL_NOTICE,
+                      "Slot export link %.40s completed export successfully. "
```

Comment:
We don't attach human readable node names anywhere. I wonder if we should make a change so that the human readable and normal name are put in the same string so that it's consistent everywhere. The current syntax of `<40 hex id>(<other name>)` is a bit clunky.